Gui: fix file dialog type filters being too long (fixes #23139)

ElementW opened the pull request:
#23139 shines the light on open/import/save/export dialog file type filters being too long, especially the wildcard `Supported formats` entry.

This PR rectifies this at the `FileDialog::get{Open,Save}FileName{,s}()` level to handle most cases, although parts of FC's code do not call this and use `QFileDialog` directly.

Uses the `QFileDialog::HideNameFilterDetails` flag to hide the actual extension filter lists, and runs all filters through a `formatFilter()` function to:
1. still show the extension list if it is deemed short enough (currently <= 128 chars in its original form, feedback welcome on that threshold)
2. deduplicate the extensions, as many have both lowercase and uppercase variants (only relevant on case-sensitive platforms)
   * this dedup is not flawless as it (as of right now anyway) does linear iteration over the extension filters and expects differing cases of the same ext to be contiguous, but most filter strings comply with this requirement anyway

The `FileDialog` functions used to force-unset the `HideNameFilterDetails` flag, I'm not quite sure why as AFAIU this flag is never set by default, nor used in FC. Setting it smoothes out platform differences [as outlined in a previous comment of mine](https://github.com/FreeCAD/FreeCAD/pull/23183#issuecomment-3199927048), and allows for this PR's behaviour.

## Issues
* #23139

This PR conflicts with #23183 on the account of solving the same issue.

## Before and After Images

<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/f6aae637-eb3b-42e5-bb90-f12c747e65e2" />

<img width="669" height="1079" alt="image" src="https://github.com/user-attachments/assets/e5348c23-6de3-4424-b90f-c7499c0b8530" />

ElementW committed:
Gui+Draft+Fem: reorder extension filters to be more consistent

In the rest of FC's code base, when multiple cases for the same file
extension exist, they are written consecutively.
Gui, Draft, and Fem have filters where this is not the case; this commit
reorders them.
Should have no side effects (other than dedup behaviour in #23209).

ElementW commented:
Argh, Qt 5 doesn't have `QStringView::slice()`, fixed that.

xtemp09 commented:
What is the purpose of `std::unique_ptr<QString>`?

ElementW committed:
Gui: deprecate open/save with flat string filter list

Replaces uses of these deprecated variant with the QStringList variant.
Does not fix the broken i18n at those callsites.
TechDraw: Fix filterList arg number

ElementW commented:
* Update to use `QStringLiteral` where relevant.
* Since this commit series adds `FileDialog::get{Open,Save}FileName{,s}(` variants with a `QStringList` filter list, add a commit to deprecate the flat list version, and fixup call sites using this older variant.

Connor9220 commented:
Need to add support for *.fctb files.  These are FreeCAD Toolbit Files that are used in the CAM workbench.  (They're the same as FCStd but with a different extension to make it easier to distinguish they're use.)

ElementW commented:
@Connor9220  Adding format associations is not within the scope of this PR (at most only touches existing import/export strings to fix inconsistencies). If you want to add that to the general open/import/save/export dialogs, `FreeCAD.addImportType()` should be added in the scripts.

Plus, [the CAM WB uses QFileDialog directly](https://github.com/FreeCAD/FreeCAD/blob/e2346dabd6080aafe0622a80524067eafa67f8f1/src/Mod/CAM/Path/Tool/assets/ui/filedialog.py#L85) so that's outside of my control anyway. I'm already working on a follow-up PR to improve on the same file dialogs as here, and once done I may look into unifying the APIs the WBs uses, but not now.

ElementW commented:
Rebased on latest `main`.

MisterMakerNL commented:
Just a question, about the image formats, have you split off the vector files? 
Because otherwise I think it could be confusing since some vector files format are not images, like DXF.

chennes commented:
Running this on Windows 11, I get the following error in the Report View
```
QWindowsNativeFileDialogBase::selectNameFilter: Invalid parameter 'Supported formats ( *.3DS *.3MF *.3ds *.3mf *.ASC *.ASMT *.AST *.BDF *.BMP *.BMS *.BREP *.BRP *.CNC *.CSG *.CSV *.CUR *.DAE *.DAT *.DWG *.DXF *.E57 *.EMN *.FCMacro *.FCMat *.FCScript *.FRD *.GC *.GCAD *.GCODE *.GIF *.GLB *.GLTF *.ICNS *.ICO *.IGES *.IGS *.INP *.IV *.JFIF *.JPEG *.JPG *.JSON *.MED *.MESHJSON *.MESHYAML *.NC *.NCC *.NGC *.OBJ *.OCA *.OFF *.PBM *.PCD *.PGM *.PLMXML *.PLY *.PNG *.PPM *.PVTU *.SHP *.SMF *.STEP *.STL *.STP *.STPZ *.SVG *.SVGZ *.TAP *.TGA *.TIF *.TIFF *.TXT *.UNV *.VRML *.VTK *.VTU *.WBMP *.WEBP *.WRL *.WRL.GZ *.WRZ *.XBM *.XDMF *.XLSX *.XML *.XPM *.YAML *.Z88 *.asc *.asmt *.ast *.bdf *.bmp *.bms *.brep *.brp *.cnc *.csg *.csv *.cur *.dae *.dat *.dwg *.dxf *.e57 *.emn *.fcmacro *.fcscript *.frd *.gc *.gcad *.gcode *.gif *.glb *.gltf *.icns *.ico *.ifc *.iges *.igs *.inp *.iv *.jfif *.jpeg *.jpg *.json *.med *.meshjson *.meshyaml *.nc *.ncc *.ngc *.obj *.oca *.off *.pbm *.pcd *.pgm *.plmxml *.ply *.png *.ppm *.pvtu *.py *.sh3d *.shp *.smf *.step *.stl *.stp *.stpZ *.stpz *.svg *.svgz *.tap *.tga *.tif *.tiff *.txt *.unv *.vrml *.vtk *.vtm *.vtu *.wbmp *.webp *.wrl *.wrl.gz *.wrz *.xbm *.xdmf *.xlsx *.xml *.xpm *.yaml *.z88)' not found in 'Supported formats ( *.3DS *.3MF *.3ds *.3mf *.ASC *.ASMT *.AST *.BDF *.BMP *.BMS *.BREP *.BRP *.CNC *.CSG *.CSV *.CUR *.DAE *.DAT *.DWG *.DXF *.E57 *.EMN *.FCMacro *.FCMat *.FCScript *.FRD *.GC *.GCAD *.GCODE *.GIF *.GLB *.GLTF *.ICNS *.ICO *.IGES *.IGS *.INP *.IV *.JFIF *.JPEG *.JPG *.JSON *.MED *.MESHJSON *.MESHYAML *.NC *.NCC *.NGC *.OBJ *.OCA *.OFF *.PBM *.PCD *.PGM *.PLMXML *.PLY *.PNG *.PPM *.PVTU *.SHP *.SMF *.STEP *.STL *.STP *.STPZ *.SVG *.SVGZ *.TAP *.TGA *.TIF *.TIFF *.TXT *.UNV *.VRML *.VTK *.VTM *.VTU *.WBMP *.WEBP *.WRL *.WRL.GZ *.WRZ *.XBM *.XDMF *.XLSX *.XML *.XPM *.YAML *.Z88 *.asc *.asmt *.ast *.bdf *.bmp *.bms *.brep *.brp *.cnc *.csg *.csv *.cur *.dae *.dat *.dwg *.dxf *.e57 *.emn *.fcmacro *.fcscript *.frd *.gc *.gcad *.gcode *.gif *.glb *.gltf *.icns *.ico *.ifc *.iges *.igs *.inp *.iv *.jfif *.jpeg *.jpg *.json *.med *.meshjson *.meshyaml *.nc *.ncc *.ngc *.obj *.oca *.off *.pbm *.pcd *.pgm *.plmxml *.ply *.png *.ppm *.pvtu *.py *.sh3d *.shp *.smf *.step *.stl *.stp *.stpZ *.stpz *.svg *.svgz *.tap *.tga *.tif *.tiff *.txt *.unv *.vrml *.vtk *.vtm *.vtu *.wbmp *.webp *.wrl *.wrl.gz *.wrz *.xbm *.xdmf *.xlsx *.xml *.xpm *.yaml *.z88), 3D Manufacturing Format (*.3mf) (*.3mf *.3MF), 3D Studio mesh (*.3ds) (*.3ds *.3DS), Alias Mesh (*.obj) (*.obj *.OBJ), Assembly Format (*.asmt) (*.asmt *.ASMT), Autodesk DWG 2D (*.dwg) (*.dwg *.DWG), Autodesk DXF 2D (*.dxf) (*.dxf *.DXF), BREP format (*.brep *.brp) (*.brep *.BREP *.brp *.BRP), Binary Mesh (*.bms) (*.bms *.BMS), Collada (*.dae) (*.dae *.DAE), Common airfoil data (*.dat) (*.dat *.DAT), Excel spreadsheet (*.xlsx) (*.xlsx *.XLSX), FEM mesh Fenics (*.xml *.xdmf) (*.xml *.XML *.xdmf *.XDMF), FEM mesh YAML/JSON (*.meshyaml *.meshjson *.yaml *.json) (*.meshyaml *.MESHYAML *.meshjson *.MESHJSON *.yaml *.YAML *.json *.JSON), FEM mesh Z88 (*.txt) (*.txt *.TXT), FEM mesh formats (*.bdf *.dat *.inp *.med *.unv *.vtk *.vtu *.pvtu *.z88) (*.bdf *.BDF *.dat *.DAT *.inp *.INP *.med *.MED *.unv *.UNV *.vtk *.VTK *.vtu *.VTU *.pvtu *.PVTU *.z88 *.Z88), FEM result CalculiX (*.frd) (*.frd *.FRD), FEM result VTK (*.vtk *.vtu *.pvtu *.vtm) (*.vtk *.VTK *.vtu *.VTU *.pvtu *.PVTU *.vtm *.VTM), FEM result Z88 displacements (*.txt) (*.txt *.TXT), FreeCAD Material Card (*.FCMat) (*.FCMat), GCode (*.nc *.gc *.ncc *.ngc *.cnc *.tap *.gcode) (*.nc *.NC *.gc *.GC *.ncc *.NCC *.ngc *.NGC *.cnc *.CNC *.tap *.TAP *.gcode *.GCODE), IDF emn file File Type (*.emn) (*.emn *.EMN), IGES format (*.iges *.igs) (*.iges *.IGES *.igs *.IGS), Image formats (*.bmp *.BMP *.cur *.CUR *.gif *.GIF *.icns *.ICNS *.ico *.ICO *.jfif *.JFIF *.jpeg *.JPEG *.jpg *.JPG *.pbm *.PBM *.pgm *.PGM *.png *.PNG *.ppm *.PPM *.svg *.SVG *.svgz *.SVGZ *.tga *.TGA *.tif *.TIF *.tiff *.TIFF *.wbmp *.WBMP *.webp *.WEBP *.xbm *.XBM *.xpm *.XPM ), Industry Foundation Classes (*.ifc) (*.ifc), Inventor V2.1 (*.iv) (*.iv *.IV), Object File Format Mesh (*.off) (*.off *.OFF), Open CAD Format (*.oca *.gcad) (*.oca *.OCA *.gcad *.GCAD), OpenSCAD CSG Format (*.csg) (*.csg *.CSG), PLMXML files (*.plmxml) (*.plmxml *.PLMXML), Point formats (*.asc *.pcd *.ply *.e57) (*.asc *.ASC *.pcd *.PCD *.ply *.PLY *.e57 *.E57), Python (*.py *.FCMacro *.FCScript) (*.py *.FCMacro *.fcmacro *.FCScript *.fcscript), STEP with colors (*.step *.stp) (*.step *.STEP *.stp *.STP), STEPZ Zip File Type (*.stpZ) (*.stpZ *.stpz *.STPZ), STL Mesh (*.stl *.ast) (*.stl *.STL *.ast *.AST), SVG as geometry (*.svg) (*.svg *.SVG), Shapefile (*.shp) (*.shp *.SHP), Simple Model Format (*.smf) (*.smf *.SMF), Spreadsheet formats (*.csv) (*.csv *.CSV), Stanford Triangle Mesh (*.ply) (*.ply *.PLY), SweetHome3D (*.sh3d) (*.sh3d), VRML V2.0 (*.wrl *.vrml *.wrz *.wrl.gz) (*.wrl *.WRL *.vrml *.VRML *.wrz *.WRZ *.wrl.gz *.WRL.GZ), Wavefront OBJ - Arch module (*.obj) (*.obj *.OBJ), glTF (*.gltf *.glb) (*.gltf *.GLTF *.glb *.GLB), All files (*.*) (*.*)'.
```
And that entire string gets displayed as the first entry in the "Import" or "Open" dialog (obviously totally bonkers, presumably a bug).

ElementW commented:
@chennes Argh, I do not have a Windows machine to test on right now, I'll spin up a VM and have a look. Hopefully the error is due to some actually invalid string from my end and not Qt's inability to parse filter strings with multiple `(...)` per entry. Failing that, falling back to Qt built-in dialogs is still an option, although subpar.
(inbefore I start rewriting the entire native dialog handling...)

ElementW commented:
I won't start drastic things if I dont need to. I'm mostly on Telegram but also occasionally on Discord.

ElementW committed:
Gui: fix file dialog type filters being too long (fixes #23139)

ElementW commented:
@chennes Looks like the error you have comes from FC remembering the last filter set you used, and said set came from a previous version where `*.VTM` was missing from the *Supported formats* entry. I believe this happens even before my changes, and is a problem that will persist until text-based filtering is phased out (which is what I set out to do a few months ago but, thanks ADHD, ended up leaving on the side).

chennes commented:
Here's what it looks like for me (safe mode):
<img width="1140" height="398" alt="Screenshot 2026-03-21 204437" src="https://github.com/user-attachments/assets/dc127457-cec6-4c8f-ac19-ae704522a054" />

So neither the "All supported formats" nor "Image files" is getting trimmed (I think those are the only two we'd expect to be here).

```
OS: Windows 11 build 26200
Architecture: x86_64
Version: 1.2.0dev.45858 (Git)
Build date: 2026/03/21 22:25:34
Build type: RelWithDebInfo
Branch: fix/file-dialog-long-filters
Hash: 09f809d3355366b18ef7d4dd556775cd64778e49
Python 3.14.2, Qt 6.10.1, Coin 4.0.6, Vtk 9.5.2, boost 1_90, Eigen3 5.0.1, PySide 6.10.1
shiboken 6.10.1, SMESH 7.7.1.0, xerces-c 3.2.5, IfcOpenShell 0.0.0, OCC 7.9.3
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: FreeCAD.qss/FreeCAD Light/
Navigation Style/Orbit Style/Rotation Mode: CAD/Rounded Arcball/Window center
Logical DPI/Physical DPI/Pixel Ratio: 96/107.975/1.5
```

ElementW committed:
Gui: fix Win32 forcing the display of extensions in Open/Save dialogs

`IFileDialog` forces the display of file filter extension lists when the
`HideFileExt` shell setting is off. Use `Get{Open,Save}FileNameW()`
instead.

ElementW committed:
Gui: FileDialog: reduce header dependencies

maxwxyz reviewed (approved):
Works.

Before:

<img width="1800" height="957" alt="Image" src="https://github.com/user-attachments/assets/d3e8b832-e497-40c2-8dfc-be44cea8e543" />

After:

<img width="671" height="1282" alt="Image" src="https://github.com/user-attachments/assets/4cb37f61-eaf4-4440-bb27-32a04a652596" />

kadet1090 commented:
Tested on linux too! Thank you so much for your effort and amazing blog posts about it!
